internal/sysfs/file_windows.go

```diff
@@ -47,20 +55,49 @@ func writeFd(fd uintptr, buf []byte) (int, sys.Errno) {
 }
 
 func readSocket(h uintptr, buf []byte) (int, sys.Errno) {
+	// Poll the socket to ensure that we never perform a blocking/overlapped Read.
+	//
+	// When the socket is closed by the remote peer, wsaPoll will return n=1 and
+	// errno=0, and syscall.ReadFile will return n=0 and errno=0 -- which indicates
+	// io.EOF.
+	if n, errno := wsaPoll(
+		[]pollFd{newPollFd(h, _POLLIN, 0)}, 0); !errors.Is(errno, sys.Errno(0)) {
+		return 0, sys.UnwrapOSError(errno)
+	} else if n <= 0 {
+		return 0, sys.EAGAIN
+	}
+
+	// Properly use overlapped result.
+	//
+	// If hFile was opened with FILE_FLAG_OVERLAPPED, the following conditions are in effect:
+	//  - The lpOverlapped parameter must point to a valid and unique OVERLAPPED structure,
+	//  otherwise the function can incorrectly report that the read operation is complete.
+	//  - The lpNumberOfBytesRead parameter should be set to NULL. Use the GetOverlappedResult
+	//  function to get the actual number of bytes read. If the hFile parameter is associated
+	//  with an I/O completion port, you can also get the number of bytes read by calling the
+	//  GetQueuedCompletionStatus function.
+	//
+	// We are currently skipping checking if hFile was opened with FILE_FLAG_OVERLAPPED but using
+	// both lpOverlapped and lpNumberOfBytesRead.
+
 	var overlapped syscall.Overlapped
 	var done uint32
 	errno := syscall.ReadFile(syscall.Handle(h), buf, &done, &overlapped)
-	if errno == syscall.ERROR_IO_PENDING {
-		errno = sys.EAGAIN
+	if errors.Is(errno, syscall.ERROR_IO_PENDING) {
+		nBytesTransferred, errno := getOverlappedResult(syscall.Handle(h), &overlapped, false)
```

Comment:
I wonder if this will ever happen, and what to do in this case. If we are at this point, `poll` has returned n >= 0 and `ReadFile` hopefully won't hang.

As we discovered together offline, ERROR_IO_PENDING means that ReadFile is holding a reference to `buf` and will complete the write to it once `getOverlappedResult()` (or related APIs such as `WaitForSingleObject`, `WaitForMultipleObjects`) returns success.

the `false` in `getOverlappedResult(syscall.Handle(h), &overlapped, false)` means that it will return immediately, with `true` meaning it will block until done. But if `false` and the I/O is still pending, we cannot really return, because `buf` could be still in an invalid state.

So, either we use `true` or we just remove `getOverlappedResult()` altogether. I am tempted to suggest the latter.

PS now I am also wondering if we need to really set the nonblocking state on the socket since we are simulating it anyway, but that's maybe for another time.

Reply:
What happens if two concurrent Reads on the same fed happened at once? I assume after both get pass wsaPoll, the slower one will start an asynchronous read... So I'm skeptical about dropping `getOverlappedResult`.

Reply:
Good point, either way if we call with false you are still going to potentially leave the buffer in an invalid state (maybe we should create a new buffer and then copy over if no error is returned? But then we're back to requiring a table of buffers) with true we will block. Maybe, assuming that concurrent access is rare, blocking is the simplest option, and we can improve in the future if necessary

Reply:
I agree, blocking might be a more proper behavior. I will also check if there's a proper way to cancel if it became async read, which might align with Unix's async reading better?

Reply:
ok